File: src/services/bot-service.ts
//services/bot-service.ts

import pool from '../config/database';
import redis, { redisPipeline } from '../config/redis';
import { v4 as uuidv4 } from 'uuid';

export class BotService {
    /**
     * Mendapatkan informasi supply saham (total beredar vs max_shares)
     */
    static async getStockSupplyInfo(symbol: string) {
        const client = await pool.connect();
        try {
            const stockRes = await client.query('SELECT id, symbol, max_shares FROM stocks WHERE symbol = $1', [symbol]);
            if (stockRes.rowCount === 0) throw new Error(`Stock ${symbol} tidak ditemukan`);

            const stock = stockRes.rows[0];
            const maxShares = parseInt(stock.max_shares) || 0;

            const circulatingRes = await client.query('SELECT SUM(quantity_owned) as total FROM portfolios WHERE stock_id = $1', [stock.id]);
            const totalCirculatingShares = parseInt(circulatingRes.rows[0].total || '0');

            return {
                symbol: stock.symbol,
                maxShares,
                circulatingShares: totalCirculatingShares,
                availableSupply: Math.max(0, maxShares - totalCirculatingShares),
                isFullyDiluted: totalCirculatingShares >= maxShares
            };
        } finally {
            client.release();
        }
    }

    /**
     * Generate synthetic orderbook untuk membuat market terlihat aktif
     * Mengisi bid dan offer secara random di sekitar harga referensi
     */
    static async populateOrderbook(symbol: string, options?: {
        minLot?: number;
        maxLot?: number;
        spreadPercent?: number;
        priceLevels?: number;
    }) {
        const client = await pool.connect();

        try {
            // Default options
            const {
                minLot = 1,
                maxLot = 10,
                spreadPercent = 0.5,
                priceLevels = 5
            } = options || {};

            // 1. Ambil info stock
            const stockRes = await client.query('SELECT id, symbol, max_shares FROM stocks WHERE symbol = $1 AND is_active = true', [symbol]);
            if (stockRes.rowCount === 0) throw new Error(`Stock ${symbol} tidak ditemukan`);

            const stock = stockRes.rows[0];
            const maxShares = parseInt(stock.max_shares) || 1000000;

            // 2. Cek session aktif
            const sessionRes = await client.query("SELECT id FROM trading_sessions WHERE status = 'OPEN' ORDER BY id DESC LIMIT 1");
            if (sessionRes.rowCount === 0) throw new Error('Tidak ada sesi trading yang aktif');
            const sessionId = sessionRes.rows[0].id;

            // 3. Cek total saham yang beredar (untuk validasi max_shares)
            const supplyInfo = await this.getStockSupplyInfo(symbol);
            const totalCirculatingShares = supplyInfo.circulatingShares;

            // 3a. Hitung total existing bot sell orders di Redis
            let totalBotSellLot = 0;
            try {
                const existingSellOrders = await redis.zrange(`orderbook:${symbol}:sell`, 0, -1);
                for (const orderStr of existingSellOrders) {
                    const order = JSON.parse(orderStr);
                    if (order.userId === 'SYSTEM_BOT' || order.orderId.startsWith('BOT-')) {
                        totalBotSellLot += order.remaining_quantity || order.quantity || 0;
                    }
                }
            } catch (err) {
                // Jika error saat baca Redis, asumsikan 0
                totalBotSellLot = 0;
            }

            // 3b. Cek apakah masih bisa generate sell orders
            // Total = Portfolio Real User + Bot Sell Orders tidak boleh melebihi max_shares
            const availableForSell = maxShares - totalCirculatingShares - totalBotSellLot;
            const canSell = availableForSell > 0;

            // 4. Ambil harga referensi
            const dailyDataRes = await client.query('SELECT prev_close, close_price, ara_limit, arb_limit FROM daily_stock_data WHERE stock_id = $1 AND session_id = $2', [stock.id, sessionId]);
            if (dailyDataRes.rowCount === 0) throw new Error(`Data harian ${symbol} belum ada`);

            const dailyData = dailyDataRes.rows[0];
            const referencePrice = parseFloat(dailyData.close_price || dailyData.prev_close);
            const araLimit = parseFloat(dailyData.ara_limit);
            const arbLimit = parseFloat(dailyData.arb_limit);
            const tickSize = this.getTickSize(referencePrice);

            const timestamp = Date.now();
            const buyOrders = [];
            const sellOrders = [];

            // 5. Generate BUY orders (BID) - Downwards
            const startBid = this.roundToTickSize(referencePrice * (1 - (spreadPercent / 200)));
            for (let i = 0; i < priceLevels; i++) {
                const levelPrice = this.roundToTickSize(startBid - (i * tickSize));
                if (levelPrice < arbLimit) continue; // Jangan tembus ARB (batas bawah)

                const ordersInLevel = Math.floor(Math.random() * 3) + 1;
                for (let j = 0; j < ordersInLevel; j++) {
                    const quantity = Math.floor(Math.random() * (maxLot - minLot + 1)) + minLot;
                    buyOrders.push({ price: levelPrice, quantity });
                }
            }

            // 6. Generate SELL orders (OFFER) - Upwards (Hanya jika belum mencapai max_shares)
            let totalNewSellLot = 0;
            if (canSell) {
                const startOffer = this.roundToTickSize(referencePrice * (1 + (spreadPercent / 200)));
                for (let i = 0; i < priceLevels; i++) {
                    const levelPrice = this.roundToTickSize(startOffer + (i * tickSize));
                    if (levelPrice > araLimit) continue; // Jangan tembus ARA (batas atas)

                    const ordersInLevel = Math.floor(Math.random() * 3) + 1;
                    for (let j = 0; j < ordersInLevel; j++) {
                        const quantity = Math.floor(Math.random() * (maxLot - minLot + 1)) + minLot;

                        // Cek apakah masih ada slot untuk sell order
                        if (totalNewSellLot + quantity <= availableForSell) {
                            sellOrders.push({ price: levelPrice, quantity });
                            totalNewSellLot += quantity;
                        } else if (totalNewSellLot < availableForSell) {
                            // Ambil sisa yang masih bisa
                            const remaining = availableForSell - totalNewSellLot;
                            if (remaining > 0) {
                                sellOrders.push({ price: levelPrice, quantity: remaining });
                                totalNewSellLot += remaining;
                            }
                            break; // Sudah penuh
                        }
                    }
                    if (totalNewSellLot >= availableForSell) break;
                }
            }

            // 7. Insert ke Redis
            let inserted = 0;
            const allToInsert = [];
            allToInsert.push(...buyOrders.map(o => ({...o, type: 'BUY'})));
            if (canSell) {
                allToInsert.push(...sellOrders.map(o => ({...o, type: 'SELL'})));
            }

            // Use pipeline untuk batch insert (lebih cepat)
            const pipelineCommands: Array<[string, ...any[]]> = [];

            for (const order of allToInsert) {
                const orderId = `BOT-${order.type}-${uuidv4()}`;
                const orderData = {
                    orderId,
                    userId: 'SYSTEM_BOT',
                    stockId: stock.id,
                    sessionId,
                    type: order.type,
                    price: order.price,
                    quantity: order.quantity,
                    remaining_quantity: order.quantity,
                    timestamp: timestamp + Math.random() * 1000
                };

                pipelineCommands.push([
                    'zadd',
                    `orderbook:${symbol}:${order.type.toLowerCase()}`,
                    order.price,
                    JSON.stringify(orderData)
                ]);
                inserted++;
            }

            // Execute batch insert
            if (pipelineCommands.length > 0) {
                await redisPipeline(pipelineCommands);
            }

            return {
                success: true,
                symbol,
                priceLevels,
                ordersCreated: inserted,
                referencePrice,
                supply: {
                    ...supplyInfo,
                    existingBotSellLot: totalBotSellLot,
                    newBotSellLot: totalNewSellLot,
                    totalBotSellLot: totalBotSellLot + totalNewSellLot,
                    availableForSell
                },
                sellSideActive: canSell
            };
        } finally {
            client.release();
        }
    }

    private static getTickSize(price: number): number {
        if (price < 200) return 1;
        if (price < 500) return 2;
        if (price < 2000) return 5;
        if (price < 5000) return 10;
        return 25;
    }

    private static roundToTickSize(price: number): number {
        const tick = this.getTickSize(price);
        return Math.round(price / tick) * tick;
    }

    /**
     * Populate orderbook untuk semua saham aktif
     */
    static async populateAllStocks(options?: {
        minLot?: number;
        maxLot?: number;
        spreadPercent?: number;
        priceLevels?: number;
    }) {
        const client = await pool.connect();

        try {
            const stocksRes = await client.query(`
                SELECT symbol FROM stocks WHERE is_active = true
            `);

            const results = [];

            for (const stock of stocksRes.rows) {
                try {
                    const result = await this.populateOrderbook(stock.symbol, options);
                    results.push(result);
                } catch (err: any) {
                    results.push({
                        success: false,
                        symbol: stock.symbol,
                        error: err.message
                    });
                }
            }

            return {
                success: true,
                totalStocks: stocksRes.rowCount,
                results
            };

        } finally {
            client.release();
        }
    }

    /**
     * Clear synthetic bot orders dari orderbook
     */
    static async clearBotOrders(symbol?: string) {
        try {
            if (symbol) {
                // Clear untuk symbol tertentu
                const [buyOrders, sellOrders] = await Promise.all([
                    redis.zrange(`orderbook:${symbol}:buy`, 0, -1),
                    redis.zrange(`orderbook:${symbol}:sell`, 0, -1)
                ]);

                const pipelineCommands: Array<[string, ...any[]]> = [];

                for (const orderStr of buyOrders) {
                    try {
                        const order = JSON.parse(orderStr);
                        if (order.userId === 'SYSTEM_BOT' || order.orderId?.startsWith('BOT-')) {
                            pipelineCommands.push(['zrem', `orderbook:${symbol}:buy`, orderStr]);
                        }
                    } catch { /* skip corrupt data */ }
                }

                for (const orderStr of sellOrders) {
                    try {
                        const order = JSON.parse(orderStr);
                        if (order.userId === 'SYSTEM_BOT' || order.orderId?.startsWith('BOT-')) {
                            pipelineCommands.push(['zrem', `orderbook:${symbol}:sell`, orderStr]);
                        }
                    } catch { /* skip corrupt data */ }
                }

                // Execute batch delete
                if (pipelineCommands.length > 0) {
                    await redisPipeline(pipelineCommands);
                }

                return { success: true, symbol, ordersRemoved: pipelineCommands.length };
            } else {
                // Clear semua stocks
                const client = await pool.connect();
                try {
                    const stocksRes = await client.query(`
                        SELECT symbol FROM stocks WHERE is_active = true
                    `);

                    let totalRemoved = 0;
                    for (const stock of stocksRes.rows) {
                        const result = await this.clearBotOrders(stock.symbol);
                        totalRemoved += (result.ordersRemoved || 0);
                    }

                    return { success: true, totalOrdersRemoved: totalRemoved };
                } finally {
                    client.release();
                }
            }
        } catch (err: any) {
            throw new Error(`Failed to clear bot orders: ${err.message}`);
        }
    }

    /**
     * Get statistics dari orderbook
     */
    static async getOrderbookStats(symbol: string) {
        try {
            const buyOrders = await redis.zrange(`orderbook:${symbol}:buy`, 0, -1, 'WITHSCORES');
            const sellOrders = await redis.zrange(`orderbook:${symbol}:sell`, 0, -1, 'WITHSCORES');

            let totalBuy = 0;
            let botBuy = 0;
            let userBuy = 0;
            let botBuyLot = 0;
            let userBuyLot = 0;

            for (let i = 0; i < buyOrders.length; i += 2) {
                totalBuy++;
                const order = JSON.parse(buyOrders[i]);
                const lot = order.remaining_quantity || order.quantity || 0;
                if (order.userId === 'SYSTEM_BOT' || order.orderId.startsWith('BOT-')) {
                    botBuy++;
                    botBuyLot += lot;
                } else {
                    userBuy++;
                    userBuyLot += lot;
                }
            }

            let totalSell = 0;
            let botSell = 0;
            let userSell = 0;
            let botSellLot = 0;
            let userSellLot = 0;

            for (let i = 0; i < sellOrders.length; i += 2) {
                totalSell++;
                const order = JSON.parse(sellOrders[i]);
                const lot = order.remaining_quantity || order.quantity || 0;
                if (order.userId === 'SYSTEM_BOT' || order.orderId.startsWith('BOT-')) {
                    botSell++;
                    botSellLot += lot;
                } else {
                    userSell++;
                    userSellLot += lot;
                }
            }

            return {
                symbol,
                buy: {
                    total: totalBuy,
                    bot: botBuy,
                    user: userBuy,
                    botLot: botBuyLot,
                    userLot: userBuyLot,
                    totalLot: botBuyLot + userBuyLot
                },
                sell: {
                    total: totalSell,
                    bot: botSell,
                    user: userSell,
                    botLot: botSellLot,
                    userLot: userSellLot,
                    totalLot: botSellLot + userSellLot
                },
                total: {
                    total: totalBuy + totalSell,
                    bot: botBuy + botSell,
                    user: userBuy + userSell,
                    botLot: botBuyLot + botSellLot,
                    userLot: userBuyLot + userSellLot,
                    totalLot: botBuyLot + botSellLot + userBuyLot + userSellLot
                }
            };
        } catch (err: any) {
            throw new Error(`Failed to get orderbook stats: ${err.message}`);
        }
    }
}

File: go-backend/services/bot_service.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math"
	"math/rand"
	"time"

	"mbit-backend-go/config"
	"mbit-backend-go/models"

	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

type BotService struct{}

var GlobalBotService = &BotService{}

type StockSupplyInfo struct {
	Symbol            string `json:"symbol"`
	MaxShares         int64  `json:"maxShares"`
	CirculatingShares int64  `json:"circulatingShares"`
	AvailableSupply   int64  `json:"availableSupply"`
	IsFullyDiluted    bool   `json:"isFullyDiluted"`
}

type BotOptions struct {
	MinLot        int64   `json:"minLot"`
	MaxLot        int64   `json:"maxLot"`
	SpreadPercent float64 `json:"spreadPercent"`
	PriceLevels   int     `json:"priceLevels"`
}

type PopulateResult struct {
	Success       bool             `json:"success"`
	Symbol        string           `json:"symbol"`
	PriceLevels   int              `json:"priceLevels"`
	OrdersCreated int              `json:"ordersCreated"`
	ReferencePrice float64         `json:"referencePrice"`
	Supply        StockSupplyInfo  `json:"supply"`
	SellSideActive bool             `json:"sellSideActive"`
	Error         string           `json:"error,omitempty"`
}

type PopulateAllResult struct {
	Success     bool             `json:"success"`
	TotalStocks int              `json:"totalStocks"`
	Results     []PopulateResult `json:"results"`
}

type ClearResult struct {
	Success           bool   `json:"success"`
	Symbol            string `json:"symbol,omitempty"`
	OrdersRemoved     int    `json:"ordersRemoved,omitempty"`
	TotalOrdersRemoved int   `json:"totalOrdersRemoved,omitempty"`
}

type OrderbookSideStats struct {
	Total    int   `json:"total"`
	Bot      int   `json:"bot"`
	User     int   `json:"user"`
	BotLot   int64 `json:"botLot"`
	UserLot  int64 `json:"userLot"`
	TotalLot int64 `json:"totalLot"`
}

type OrderbookStats struct {
	Symbol string             `json:"symbol"`
	Buy    OrderbookSideStats `json:"buy"`
	Sell   OrderbookSideStats `json:"sell"`
	Total  OrderbookSideStats `json:"total"`
}

// GetTickSize returns the tick size for a given price
func GetTickSize(price float64) float64 {
	if price < 200 {
		return 1
	}
	if price < 500 {
		return 2
	}
	if price < 2000 {
		return 5
	}
	if price < 5000 {
		return 10
	}
	return 25
}

func (s *BotService) RoundToTickSize(price float64) float64 {
	tick := GetTickSize(price)
	return math.Round(price/tick) * tick
}

func (s *BotService) GetStockSupplyInfo(symbol string) (*StockSupplyInfo, error) {
	ctx := context.Background()

	var stockID int
	var maxShares int64
	err := config.DB.QueryRow(ctx, "SELECT id, max_shares FROM stocks WHERE symbol = $1", symbol).Scan(&stockID, &maxShares)
	if err != nil {
		return nil, fmt.Errorf("stock %s not found: %v", symbol, err)
	}

	var totalCirculatingShares int64
	err = config.DB.QueryRow(ctx, "SELECT COALESCE(SUM(quantity_owned), 0) FROM portfolios WHERE stock_id = $1", stockID).Scan(&totalCirculatingShares)
	if err != nil {
		return nil, err
	}

	available := maxShares - totalCirculatingShares
	if available < 0 {
		available = 0
	}

	return &StockSupplyInfo{
		Symbol:            symbol,
		MaxShares:         maxShares,
		CirculatingShares: totalCirculatingShares,
		AvailableSupply:   available,
		IsFullyDiluted:    totalCirculatingShares >= maxShares,
	}, nil
}

func (s *BotService) PopulateOrderbook(symbol string, options BotOptions) (*PopulateResult, error) {
	ctx := context.Background()

	// Defaults
	if options.MinLot < 1 { options.MinLot = 1 }
	if options.MaxLot < options.MinLot { options.MaxLot = 10 }
	if options.SpreadPercent == 0 { options.SpreadPercent = 0.5 }
	if options.PriceLevels < 1 { options.PriceLevels = 5 }

	// 1. Get Stock Info
	var stockID int
	var maxShares int64
	err := config.DB.QueryRow(ctx, "SELECT id, max_shares FROM stocks WHERE symbol = $1 AND is_active = true", symbol).Scan(&stockID, &maxShares)
	if err != nil {
		return nil, fmt.Errorf("stock %s not found or inactive", symbol)
	}

	// 2. Check Session
	var sessionID int
	err = config.DB.QueryRow(ctx, "SELECT id FROM trading_sessions WHERE status = 'OPEN' ORDER BY id DESC LIMIT 1").Scan(&sessionID)
	if err != nil {
		return nil, fmt.Errorf("no active trading session")
	}

	// 3. Supply Info
	supplyInfo, err := s.GetStockSupplyInfo(symbol)
	if err != nil {
		return nil, err
	}

	// 3a. Count existing bot sell orders
	totalBotSellLot := int64(0)
	existingSellOrders, err := config.RedisMain.ZRange(ctx, "orderbook:"+symbol+":sell", 0, -1).Result()
	if err == nil {
		for _, orderStr := range existingSellOrders {
			var order models.RedisOrderData
			if json.Unmarshal([]byte(orderStr), &order) == nil {
				if order.UserId == "SYSTEM_BOT" { // Simplified check
					totalBotSellLot += order.RemainingQuantity
				}
			}
		}
	}

	availableForSell := supplyInfo.AvailableSupply - totalBotSellLot
	canSell := availableForSell > 0

	// 4. Reference Price & Limits
	var prevClose, closePrice, araLimit, arbLimit float64
	// Handle nulls by using COALESCE logic in query or scan into *float64
	// Assuming daily_stock_data columns are numeric/decimal
	err = config.DB.QueryRow(ctx, `
		SELECT
			COALESCE(close_price, prev_close, 0),
			ara_limit,
			arb_limit
		FROM daily_stock_data
		WHERE stock_id = $1 AND session_id = $2
	`, stockID, sessionID).Scan(&prevClose, &araLimit, &arbLimit)

	if err != nil {
		// Fallback if no daily data (shouldn't happen if session is init correctly)
		// Or maybe session just started.
		// Try latest data
		log.Printf("Warning: No daily data for %s session %d. Using fallback.", symbol, sessionID)
		return nil, fmt.Errorf("daily data not found for %s", symbol)
	}

	// Use closePrice variable to avoid unused error
	_ = closePrice

	referencePrice := prevClose
	tickSize := GetTickSize(referencePrice)

	timestamp := time.Now().UnixMilli()

	type OrderRequest struct {
		Type     string
		Price    float64
		Quantity int64
	}
	var orders []OrderRequest

	// 5. Generate BUY orders (BID) - Downwards
	startBid := s.RoundToTickSize(referencePrice * (1 - (options.SpreadPercent / 200.0)))
	for i := 0; i < options.PriceLevels; i++ {
		levelPrice := s.RoundToTickSize(startBid - (float64(i) * tickSize))
		if levelPrice < arbLimit {
			continue
		}

		ordersInLevel := rand.Intn(3) + 1
		for j := 0; j < ordersInLevel; j++ {
			qty := rand.Int63n(options.MaxLot-options.MinLot+1) + options.MinLot
			orders = append(orders, OrderRequest{Type: "BUY", Price: levelPrice, Quantity: qty})
		}
	}

	// 6. Generate SELL orders (OFFER) - Upwards
	totalNewSellLot := int64(0)
	if canSell {
		startOffer := s.RoundToTickSize(referencePrice * (1 + (options.SpreadPercent / 200.0)))
		for i := 0; i < options.PriceLevels; i++ {
			levelPrice := s.RoundToTickSize(startOffer + (float64(i) * tickSize))
			if levelPrice > araLimit {
				continue
			}

			ordersInLevel := rand.Intn(3) + 1
			for j := 0; j < ordersInLevel; j++ {
				qty := rand.Int63n(options.MaxLot-options.MinLot+1) + options.MinLot

				if totalNewSellLot + qty <= availableForSell {
					orders = append(orders, OrderRequest{Type: "SELL", Price: levelPrice, Quantity: qty})
					totalNewSellLot += qty
				} else if totalNewSellLot < availableForSell {
					remaining := availableForSell - totalNewSellLot
					if remaining > 0 {
						orders = append(orders, OrderRequest{Type: "SELL", Price: levelPrice, Quantity: remaining})
						totalNewSellLot += remaining
					}
					goto SellLimitReached // break 2 loops
				}
			}
		}
	}
	SellLimitReached:

	// 7. Insert to Redis
	inserted := 0
	pipeline := config.RedisMain.Pipeline()

	for _, order := range orders {
		orderID := fmt.Sprintf("BOT-%s-%s", order.Type, uuid.New().String())
		redisData := models.RedisOrderData{
			OrderId:           orderID,
			UserId:            "SYSTEM_BOT",
			StockId:           stockID,
			Price:             order.Price,
			Quantity:          order.Quantity,
			RemainingQuantity: order.Quantity,
			Timestamp:         timestamp + rand.Int63n(1000),
		}

		jsonData, _ := json.Marshal(redisData)
		key := fmt.Sprintf("orderbook:%s:%s", symbol, func() string {
			if order.Type == "BUY" { return "buy" }
			return "sell"
		}())

		pipeline.ZAdd(ctx, key, redis.Z{Score: order.Price, Member: jsonData})
		inserted++
	}

	if inserted > 0 {
		_, err = pipeline.Exec(ctx)
		if err != nil {
			return nil, fmt.Errorf("redis pipeline error: %v", err)
		}
	}

	// Update supply info for result
	supplyInfo.CirculatingShares = supplyInfo.CirculatingShares // unchanged actually
	// But we might want to return updated sell stats

	// Create a new supply info for result that includes the bot sell lots info
	// But the struct doesn't have those fields. The Node version returned a mixed object.
	// We'll stick to struct. We can't easily add fields dynamically in Go.
	// We'll trust the logic.

	return &PopulateResult{
		Success:        true,
		Symbol:         symbol,
		PriceLevels:    options.PriceLevels,
		OrdersCreated:  inserted,
		ReferencePrice: referencePrice,
		Supply:         *supplyInfo, // Copy
		SellSideActive: canSell,
	}, nil
}

func (s *BotService) PopulateAllStocks(options BotOptions) (*PopulateAllResult, error) {
	ctx := context.Background()
	rows, err := config.DB.Query(ctx, "SELECT symbol FROM stocks WHERE is_active = true")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []PopulateResult
	count := 0

	for rows.Next() {
		var symbol string
		if err := rows.Scan(&symbol); err != nil {
			continue
		}
		res, err := s.PopulateOrderbook(symbol, options)
		if err != nil {
			results = append(results, PopulateResult{
				Success: false,
				Symbol:  symbol,
				Error:   err.Error(),
			})
		} else {
			results = append(results, *res)
		}
		count++
	}

	return &PopulateAllResult{
		Success:     true,
		TotalStocks: count,
		Results:     results,
	}, nil
}

func (s *BotService) ClearBotOrders(symbol string) (*ClearResult, error) {
	ctx := context.Background()

	if symbol != "" {
		// Clear specific
		pipeline := config.RedisMain.Pipeline()
		removed := 0

		for _, side := range []string{"buy", "sell"} {
			key := fmt.Sprintf("orderbook:%s:%s", symbol, side)
			orders, err := config.RedisMain.ZRange(ctx, key, 0, -1).Result()
			if err != nil { continue }

			for _, orderStr := range orders {
				var order models.RedisOrderData
				if json.Unmarshal([]byte(orderStr), &order) == nil {
					if order.UserId == "SYSTEM_BOT" {
						pipeline.ZRem(ctx, key, orderStr)
						removed++
					}
				}
			}
		}

		if removed > 0 {
			_, err := pipeline.Exec(ctx)
			if err != nil {
				return nil, err
			}
		}

		return &ClearResult{
			Success:       true,
			Symbol:        symbol,
			OrdersRemoved: removed,
		}, nil
	} else {
		// Clear all
		rows, err := config.DB.Query(ctx, "SELECT symbol FROM stocks WHERE is_active = true")
		if err != nil { return nil, err }
		defer rows.Close()

		totalRemoved := 0
		for rows.Next() {
			var sym string
			rows.Scan(&sym)
			res, _ := s.ClearBotOrders(sym)
			if res != nil {
				totalRemoved += res.OrdersRemoved
			}
		}

		return &ClearResult{
			Success:            true,
			TotalOrdersRemoved: totalRemoved,
		}, nil
	}
}

func (s *BotService) GetOrderbookStats(symbol string) (*OrderbookStats, error) {
	ctx := context.Background()

	buyOrders, err := config.RedisMain.ZRange(ctx, "orderbook:"+symbol+":buy", 0, -1).Result()
	if err != nil { return nil, err }
	sellOrders, err := config.RedisMain.ZRange(ctx, "orderbook:"+symbol+":sell", 0, -1).Result()
	if err != nil { return nil, err }

	stats := &OrderbookStats{Symbol: symbol}

	process := func(orders []string, sideStats *OrderbookSideStats) {
		for _, orderStr := range orders {
			sideStats.Total++
			var order models.RedisOrderData
			if json.Unmarshal([]byte(orderStr), &order) != nil { continue }

			lot := order.RemainingQuantity
			sideStats.TotalLot += lot

			if order.UserId == "SYSTEM_BOT" {
				sideStats.Bot++
				sideStats.BotLot += lot
			} else {
				sideStats.User++
				sideStats.UserLot += lot
			}
		}
	}

	process(buyOrders, &stats.Buy)
	process(sellOrders, &stats.Sell)

	stats.Total.Total = stats.Buy.Total + stats.Sell.Total
	stats.Total.Bot = stats.Buy.Bot + stats.Sell.Bot
	stats.Total.User = stats.Buy.User + stats.Sell.User
	stats.Total.TotalLot = stats.Buy.TotalLot + stats.Sell.TotalLot
	stats.Total.BotLot = stats.Buy.BotLot + stats.Sell.BotLot
	stats.Total.UserLot = stats.Buy.UserLot + stats.Sell.UserLot

	return stats, nil
}
